File: api/columns.py
from api._lib.blob_io import load_df
from api._lib.core.feed_processor import (
    RECOMMENDED_RAW,
    apply_column_selection,
    default_keep_map,
)
from api._lib.core.utils import norm
from api._lib.handler_utils import BaseHandler, clicks_sorted_preview, column_meta, records


class handler(BaseHandler):
    def do_POST(self):
        try:
            body = self._json_body()
            raw_df = load_df(body["rawDfUrl"])
            action = body.get("action")

            if action == "recommended":
                n2o = {norm(c): c for c in raw_df.columns}
                keep_map = {c: False for c in raw_df.columns}
                for item in {norm(x) for x in RECOMMENDED_RAW}:
                    if item in n2o:
                        keep_map[n2o[item]] = True
            elif action == "all":
                keep_map = {c: True for c in raw_df.columns}
            elif action == "none":
                keep_map = {c: False for c in raw_df.columns}
            elif action == "invert":
                old = body.get("keepMap", {})
                keep_map = {c: not old.get(c, False) for c in raw_df.columns}
            else:
                keep_map = body.get("keepMap", default_keep_map(raw_df))

            kept = apply_column_selection(raw_df, keep_map)
            self._send_json(200, {
                "keepMap": keep_map,
                "columns": column_meta(raw_df, keep_map),
                "preview": records(clicks_sorted_preview(raw_df, kept)),
                "rowCount": len(kept),
                "colCount": len(kept.columns),
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/columns.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import { sendError } from '../api/_lib/handlerUtils'
import { applyColumnSelection, defaultKeepMap, countProducts } from '../api/_lib/core/feedProcessor'
import { columnMeta, clicksSortedPreview } from '../api/_lib/handlerUtils'
import { loadRows } from '../api/_lib/sessionStore'

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' })

  try {
    const { rawDfUrl, keepMap: inKeepMap, action } = req.body as {
      rawDfUrl: string
      keepMap?: Record<string, boolean>
      action?: 'recommended' | 'all' | 'none' | 'invert'
    }

    const rows = await loadRows(rawDfUrl)
    const cols = Object.keys(rows[0] ?? {})
    let keepMap = inKeepMap ?? defaultKeepMap(cols)

    if (action === 'recommended') keepMap = defaultKeepMap(cols)
    else if (action === 'all') keepMap = Object.fromEntries(cols.map(c => [c, true]))
    else if (action === 'none') keepMap = Object.fromEntries(cols.map(c => [c, false]))
    else if (action === 'invert') keepMap = Object.fromEntries(cols.map(c => [c, !keepMap[c]]))

    const preview = clicksSortedPreview(applyColumnSelection(rows, keepMap))

    res.status(200).json({
      keepMap,
      columns: columnMeta(rows, keepMap),
      preview,
      rowCount: rows.length,
      colCount: cols.filter(c => keepMap[c]).length,
    })
  } catch (err) {
    sendError(res, err)
  }
}
